src/librustc_mir/transform/lower_intrinsics.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Some intrinsics can be lowered to MIR code. Doing this enables
+//! more optimizations to happen.
+
+use std::borrow::Cow;
+
+use rustc::middle::const_val::ConstVal;
+use rustc::mir::*;
+use rustc::mir::transform::{MirPass, MirSource, Pass};
+use rustc::ty::{TyCtxt, TypeVariants};
+use syntax::abi::Abi;
+use syntax_pos::symbol::InternedString;
+
+pub struct LowerIntrinsics;
+
+impl Pass for LowerIntrinsics {
+    fn name(&self) -> Cow<'static, str> { "LowerIntrinsics".into() }
+}
+
+impl<'tcx> MirPass<'tcx> for LowerIntrinsics {
+    fn run_pass<'a>(&mut self, tcx: TyCtxt<'a, 'tcx, 'tcx>, _: MirSource, mir: &mut Mir<'tcx>) {
+        for bb_data in mir.basic_blocks_mut() {
+            if let Some((func, args, target)) = intrinsic_call(bb_data) {
+                lower_intrinsic(tcx, bb_data, func, args, target);
+            }
+        }
+    }
+}
+
+// Returns the elements of the function call in case it is an intrinsic
+fn intrinsic_call<'tcx>(bb_data: &BasicBlockData<'tcx>)
+                        -> Option<(Constant<'tcx>, Vec<Operand<'tcx>>, BasicBlock)> {
+    use self::TerminatorKind::Call;
+    match bb_data.terminator().kind {
+        Call { func: Operand::Constant(ref func), ref args, ref destination, .. } => {
+            // Note: for all intrinsics that we lower in this pass, we assume that they
+            // don't require cleanup. If an intrinsic happens to require cleanup,
+            // the cleanup code will be happily ignored and bad things will happen.
+
+            match func.ty.sty {
+                TypeVariants::TyFnDef(_, _, fn_sig)
```

Comment:
Well, I needed to break the line somewhere to avoid the 100 width limit. Maybe I could indent the `if` statement. I haven't seen a similar situation earlier, so I just went with what felt natural to me.